### PR TITLE
Deutsch-Jozsa

### LINKED ISSUE #96: Clean up Deutsch-Jozsa
Deutsch-Jozsa need to be cleaned up and moved out of alpha.

### PR DESCRIPTION
**Pull request by ampolloreno**
Cleaning up Deutsch-Jozsa. Fixes #96

### CODE REVIEW

File: docs/deutsch_jozsa.rst [hunk 1]
```diff
@@ -5,11 +5,48 @@ Deutsch-Jozsa Algorithm
...
-to a single bit is constant or balanced, provided that it is one of the two. A constant
-function always maps to either 1 or 0, and a balanced function maps to 1 for half of the inputs and maps to 0
-for the other half. Unlike any classical algorithm, the Deutsch-Jozsa Algorithm can solve this problem with
-a single iteration, regardless of the input size.
+The Deutsch-Jozsa algorithm can determine whether a function mapping all bitstrings to a single bit
+is constant or balanced, provided that it is one of the two. A constant function always maps to
+either 1 or 0, and a balanced function maps to 1 for half of the inputs and maps to 0 for the other
+half. Unlike any deterministic classical algorithm, the Deutsch-Jozsa Algorithm can solve this
+problem with a single iteration, regardless of the input size. It was one of the first known quantum
+algorithms that showed an exponential speedup, albeit it against a deterministic (non-probabilistic)
```
Comment (mpharrigan) [inline docs/deutsch_jozsa.rst]: "albeit it"

Comment (ampolloreno) [reply docs/deutsch_jozsa.rst]: great

File: grove/deutsch_jozsa/deutsch_jozsa.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        :param mappings: Dictionary of the mappings of f(x) on all length n bitstrings, e.g.
+
+            >>> {'00': '0', '01': '1', '10': '1', '11': '0'}
+
+        :type mappings: Dict[String, Int]
+        :return: ndarray representing specified unitary transformation.
+        :rtype: np.ndarray
+        """
+        num_qubits = int(np.log2(len(mappings)))
+        bitsum = sum([int(bit) for bit in mappings.values()])
```
Comment (jotterbach) [inline grove/deutsch_jozsa/deutsch_jozsa.py]: this is interesting! Now we need a loop over the full function! This is O(2**n) which is the classic system's runtime. Are you implying we cannot construct the oracle faster than the classic algorithm runtime? If so then what's the point of DJ? We should highlight this!

File: grove/deutsch_jozsa/deutsch_jozsa.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                unitary_funct[i, j] = 1
+            return np.kron(np.identity(2), unitary_funct)
+
+        # Only ones were entered
+        elif bitsum == 2 ** num_qubits:
+            x_gate = np.array([[0, 1], [1, 0]])
+            return np.kron(SWAP_MATRIX, np.identity(2 ** (num_qubits - 1))).dot(
+                np.kron(x_gate, np.identity(2 ** num_qubits)))
+        else:
+            raise ValueError("f(x) must be constant or balanced")
```
Comment (jotterbach) [inline grove/deutsch_jozsa/deutsch_jozsa.py]: You just wrapped the classic algorithm to check the solution to generate the oracle. This renders the quantum algorithm fairly useless? Am I missing something?

File: grove/deutsch_jozsa/deutsch_jozsa.py (hunk 2 above)
Comment (ampolloreno) [reply grove/deutsch_jozsa/deutsch_jozsa.py]: So this implementation (written over the summer) constructs an oracle for a given function. What part specifically are you saying is 2**n? Certainly we're `defgate`- ing the oracle, and that matrix is exponentially large.

Normally, in any of these cases, you're handed an oracle. We're explicitly defining example oracles, and so to do that in full generality it seems to be that this is going to require this kind of scaling, right?

File: grove/deutsch_jozsa/deutsch_jozsa.py (hunk 3 above)
Comment (ampolloreno) [reply grove/deutsch_jozsa/deutsch_jozsa.py]: Can you elaborate on what you mean? The function is provided as a bitmap, and the values are iterated over, and the matrix is built. What do you mean by "check the solution to generate the oracle"?

Comment (jotterbach) [reply grove/deutsch_jozsa/deutsch_jozsa.py]: Well, what I mean by this, is that to generate the oracle you walk over the full bitmap, so you can find the solution by just checking every element.

File: grove/deutsch_jozsa/deutsch_jozsa.py (hunk 2 above)
Comment (jotterbach) [reply grove/deutsch_jozsa/deutsch_jozsa.py]: yes, the def-gate is exponentially large, but it's also super-sparse. do the non-sparse elements also scale exponentially in their generation? If yes, how do other problems get around this?

**Comment by jotterbach** [comment]
@ampolloreno the PR has merge conflicts

File: grove/deutsch_jozsa/deutsch_jozsa.py (hunk 3 above)
Comment (ampolloreno) [reply grove/deutsch_jozsa/deutsch_jozsa.py]: I'll address this in documentation.

File: grove/deutsch_jozsa/deutsch_jozsa.py (hunk 2 above)
Comment (ampolloreno) [reply grove/deutsch_jozsa/deutsch_jozsa.py]: I'l address this in documentation.

**Pull request merged**